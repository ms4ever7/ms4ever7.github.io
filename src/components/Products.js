import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { fetchProducts, setPrimaryItem } from '../redux/actions/products';

import NoResultsFound from './NoResultsFound';
import Product from './Product';

class Products extends Component {
  static propTypes = {
    products: PropTypes.array,
    selectedCategories: PropTypes.array,
    selectedColors: PropTypes.array,
    fetchProducts: PropTypes.func,
    setPrimaryItem: PropTypes.func
  }

  static defaultProps = {
    products: [],
    selectedCategories: [],
    selectedColors: [],
    fetchProducts: () => {}
  }

  componentDidMount() {
    this.props.fetchProducts();
  }

  get appropriateItems() {
    const { products, selectedCategories, selectedColors } = this.props;

    return products.filter((product) => {
      const { category, colors } = product;

      if ((selectedCategories.length && !selectedCategories.includes(category))
        || (selectedColors.length && !selectedColors.every(color => colors.includes(color)))) {
        return;
      }

      return product;
    });
  }


  get items() {
    const { selectedCategories, selectedColors, setPrimaryItem } = this.props;

    if (!this.appropriateItems.length) {
      return <NoResultsFound selectedCategories={selectedCategories} selectedColors={selectedColors} />;
    }

    return this.appropriateItems.map(item => <Product key={item.id} data={item} setPrimaryItem={setPrimaryItem} className='productItem' />);
  }

  render() {
    const { products } = this.props;

    if (!products || !products.length) {
      return null;
    }

    return (
      <div className='products__wrapper layout layout-align-center-start flex'>
        {this.items}
      </div>
    );
  }
}

const mapStateToProps = ({ products, filters }) => ({
  products: products.items,
  selectedCategories: filters.selectedCategories || [],
  selectedColors: filters.selectedColors || []
});

const mapDispatchToProps = dispatch => ({
  fetchProducts: () => dispatch(fetchProducts()),
  setPrimaryItem: (item, id) => dispatch(setPrimaryItem(item, id))
});

export default connect(mapStateToProps, mapDispatchToProps)(Products);
