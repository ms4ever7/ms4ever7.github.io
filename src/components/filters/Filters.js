import React, { Component } from 'react';
import { connect } from 'react-redux';

import Categories from './Categories';
import Colors from './Colors';

import { resetFilters } from '../../redux/actions/filters';

class Filters extends Component {
  render() {
    return (
      <div className='filters__wrapper layout-column layout-align-start-start'>
        <Categories />
        <Colors />
        <div className='filters__reset-button' onClick={this.props.resetFilters}>Reset all filters</div>
      </div>
    );
  }
}

const mapStateToProps = () => ({});

const mapDispatchToProps = dispatch => ({
  resetFilters: () => dispatch(resetFilters())
});


export default connect(mapStateToProps, mapDispatchToProps)(Filters);