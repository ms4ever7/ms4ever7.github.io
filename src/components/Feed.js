
import React, { Component } from 'react';
import { connect } from 'react-redux';

import Filters from './filters/Filters';
import Products from './Products';

class Feed extends Component {
  render() {
    return (
      <div className='feed__container layout layout-align-center-start'>
        <Filters />
        <Products />
      </div>
    );
  }
}


export default Feed;