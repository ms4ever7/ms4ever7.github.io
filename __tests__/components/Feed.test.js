import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Feed from '../../src/components/Feed';
import Filters from '../../src/components/filters/Filters';
import Products from '../../src/components/Products';

describe('Feed', () => {
  let wrapper;


  it('should render correctly', () => {
    wrapper = shallow(
      <Feed />
    );

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('has appropriate classname', () => {
    const wrapper = shallow(<Feed />);

    expect(wrapper.contains(<Filters />)).toBe(true);
    expect(wrapper.contains(<Products />)).toBe(true);
  });
});
