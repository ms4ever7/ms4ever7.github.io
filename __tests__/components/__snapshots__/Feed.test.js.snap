// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feed should render correctly 1`] = `
<div
  className="feed__container layout layout-align-center-start"
>
  <Connect(Filters) />
  <Connect(Products) />
</div>
`;
